src/library_manager/lib_manager.c

```diff
@@ -68,114 +68,133 @@ static int lib_manager_load_data_from_storage(void __sparse_cache *vma, void *s_
 	return 0;
 }
 
-static int lib_manager_load_module(uint32_t module_id, struct sof_man_module *mod,
-				   struct sof_man_fw_desc *desc)
+static int lib_manager_load_module(const uint32_t module_id,
+				   const struct sof_man_module *const mod)
 {
-	struct ext_library *ext_lib = ext_lib_get();
-	uint32_t lib_id = LIB_MANAGER_GET_LIB_ID(module_id);
-	size_t load_offset = (size_t)((void *)ext_lib->desc[lib_id]);
-	void __sparse_cache *va_base_text = (void __sparse_cache *)
-		mod->segment[SOF_MAN_SEGMENT_TEXT].v_base_addr;
-	void *src_txt = (void *)(mod->segment[SOF_MAN_SEGMENT_TEXT].file_offset + load_offset);
-	size_t st_text_size = mod->segment[SOF_MAN_SEGMENT_TEXT].flags.r.length;
-	void __sparse_cache *va_base_rodata = (void __sparse_cache *)
-		mod->segment[SOF_MAN_SEGMENT_RODATA].v_base_addr;
-	void *src_rodata =
-		(void *)(mod->segment[SOF_MAN_SEGMENT_RODATA].file_offset + load_offset);
-	size_t st_rodata_size = mod->segment[SOF_MAN_SEGMENT_RODATA].flags.r.length;
+	const struct ext_library *const ext_lib = ext_lib_get();
+	const uint32_t lib_id = LIB_MANAGER_GET_LIB_ID(module_id);
+	const uintptr_t load_offset = POINTER_TO_UINT(ext_lib->desc[lib_id]);
+	void *src, *va_base;
+	size_t size;
+	uint32_t flags;
+	int ret, idx;
+
+	for (idx = 0; idx < ARRAY_SIZE(mod->segment); ++idx) {
+		if (!mod->segment[idx].flags.r.load)
+			continue;
```

Comment:
this skips .bss, so effectively you made a loop for 2 elements. Should we go a step further and integrate .bss here too by either copying data for .data and .text or clearing the buffer for .bss?

Reply:
@lyakh The .bss section is handled differently. Its size is divided by the maximum number of module instances and only the selected fragment for a given instance is mapped.

Reply:
well, that special .bss handling seems strange to me - it's too ad-hoc IMHO. This means, that you cannot have the same file used both as a loadable module and built-in, because its .bss will be handled differently. Besides it's too special - developers would need to think about it every time they put something in .bss. We discussed this with @pjdobrowolski briefly and maybe we can and should drop that behaviour.
And if we don't and we keep .bss out and only handle .text and .data here, we effectively get a loop over three elements to then throw away one of them, leaving a loop over two iterations - you could just keep it as is unfolded.

Reply:
@softwarecki @lyakh any update here, should we put a big inline comment here spelling out this difference and its expected behavior from ROM i.e. its done for current compatibility reasons

Reply:
any communication between modules, build-in or loadable should be done by using native-system-services. I'm no sure if making shared bss is good practice for such behavior. Especially if we load and unload modules on fly, what will happen with common bss? Each module will leave it own stuff in there until we will have an overflow/

Reply:
@pjdobrowolski what do you mean by "shared bss?" The "normal" programmer doesn't care about ELF sections. Those are just executable format details. They just put `static int x;` at the file level and they understand that there will be exactly one instance of that `x` on the entire system. The size of that `x` would usually be 4 bytes. If your module specifies maximum instance count of 8, you end up with half a byte of that `x` per instance. What is that good for? If the programmer wants to make per-instance data, they make `static int x[max_instance_count]` and then they count instances in their instantiation routine and use the respective member of the array. How are you going to use a "part of .bss" in your software?

Reply:
DSP has restricted 3MB of ram which is shared between modules. BSS is defining how much of we use. Can you explain me how removing BSS sections from elf will improve working on such small space?

Reply:
@lyakh You wrote: "This means, that you cannot have the same file used both as a loadable module and built-in...". Where you see use case to have the same code used as loadable and built-in at the same time? I think we do loadable to not have given module built into base FW all the time. 
Another think, there are other reasons for having separate .bss section for each module. 
One is backward compatibility with reference FW design.
The other even more important is possibility to use in the feature some memory protection features. Mixing up the module .bss with base FW general .bss section will block us from introducing such solution. 
The third reason is the debuggability of base FW with presence of external modules. If we mix up the memory spaces, tracking any failures will be more difficult since external modules could corrupt base FW data.

Reply:
@pjdobrowolski @jxstelter I think there's some confusion here. I'm not proposing to remove .bss or to share .bss *between modules* or anything special of that kind. I just propose to do what C does: you program
```
static int module_x;
```
and it's placed in .bss and it's available there to that *file* (because it's `static`). Or you have a global
```
int global_x;
```
and it's also in .bss and accessible globally. In all cases there's exactly 1 copy of that object. What I don't understand why loadable modules are linking those objects with module *instances*. That means, that e.g. if you make volume a loadable module. And you activate 3 pipelines and each of them has a volume component in it so you have 3 instances. So then you give to each *instance* i.e. to each volume component in each pipeline a separate *part* of that .bss, i.e. a part of those objects. I understand *why* it is done, but I don't find this a good idea. It is done to "emulate" a poor-man allocation, so you want to put something like
```
static struct module_private_data[MAX_MODULE_INSTANCES];
```
in those modules and then give each instance one of those objects. But I don't find that a good approach. I think it's confusing and misinterpreting the meaning of the ELF .bss section

Reply:
If you would look more carefully in history of commits, you would find out, that connecting .bss with instances was introduce much before idea of "poor-man allocation".

Reply:
ok, so I think we probably need to make relationship between bss and instances more visible with inline comments in the code here, this way we can avoid any misunderstandings as we continue and add logic for services API and IADK modules.